src/btree/bt_page.c

```diff
@@ -584,72 +588,63 @@ __inmem_row_leaf(WT_SESSION_IMPL *session, WT_PAGE *page)
-
-        /* Allocate the per-page update array if one doesn't already exist. */
-        if (page->entries != 0 && page->modify->mod_row_update == NULL)
-            WT_PAGE_ALLOC_AND_SWAP(
-              session, page, page->modify->mod_row_update, upd_array, page->entries);
-
-        /* For each entry in the page */
-        size = total_size = 0;
-        upd_array = page->modify->mod_row_update;
-        WT_ROW_FOREACH (page, rip, i) {
-            /* Unpack the on-page value cell. */
-            __wt_row_leaf_value_cell(session, page, rip, NULL, &unpack);
-            if (unpack.tw.prepare) {
-                /* Take the value from the original page cell. */
-                if (value == NULL)
-                    WT_ERR(__wt_scr_alloc(session, 0, &value));
-                WT_ERR(__wt_page_cell_data_ref(session, page, &unpack, value));
-
-                WT_ERR(__wt_upd_alloc(session, value, WT_UPDATE_STANDARD, &upd, &size));
-                total_size += size;
-                upd->durable_ts = unpack.tw.durable_start_ts;
-                upd->start_ts = unpack.tw.start_ts;
-                upd->txnid = unpack.tw.start_txn;
-
-                /*
-                 * Instantiating both update and tombstone if the prepared update is a tombstone.
-                 * This is required to ensure that written prepared delete operation must be removed
-                 * from the data store, when the prepared transaction gets rollback.
-                 */
-                if (WT_TIME_WINDOW_HAS_STOP(&unpack.tw)) {
-                    WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, &size));
-                    total_size += size;
-                    tombstone->durable_ts = WT_TS_NONE;
-                    tombstone->start_ts = unpack.tw.stop_ts;
-                    tombstone->txnid = unpack.tw.stop_txn;
-                    tombstone->prepare_state = WT_PREPARE_INPROGRESS;
-                    F_SET(tombstone, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
-                    F_SET(upd, WT_UPDATE_RESTORED_FROM_DS);
-                    tombstone->next = upd;
-                } else {
-                    upd->durable_ts = WT_TS_NONE;
-                    upd->prepare_state = WT_PREPARE_INPROGRESS;
-                    F_SET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
-                    tombstone = upd;
-                }
-
-                upd_array[WT_ROW_SLOT(page, rip)] = tombstone;
-                tombstone = upd = NULL;
-            }
+        if (!unpack.tw.prepare || prepare == PREPARE_IGNORE)
+            continue;
+
+        /* First prepared transaction setup. */
+        if (prepare == PREPARE_INSTANTIATE) {
+            WT_ERR(__wt_page_modify_init(session, page));
+            if (!F_ISSET(btree, WT_BTREE_READONLY))
+                __wt_page_modify_set(session, page);
+
+            /* Allocate the per-page update array. */
+            WT_ERR(__wt_calloc_def(session, page->entries, &page->modify->mod_row_update));
```

Comment:
I think the memory allocation for `mod_row_update` is not considered as a page cache size. we can increase the `total_size` with that allocation memory.

Reply:
Thanks, @kommiharibabu, fixed in 5ddec36c9.